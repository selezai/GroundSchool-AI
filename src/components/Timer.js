import React, { useState, useEffect, useCallback } from 'react';
import { View, Text, StyleSheet } from 'react-native';
import PropTypes from 'prop-types';

const Timer = ({
  duration,
  onTimeUp,
  isRunning = true,
  showHours = false,
  testID = 'timer'
}) => {
  const [timeLeft, setTimeLeft] = useState(duration);

  const formatTime = useCallback((time) => {
    const hours = Math.floor(time / 3600);
    const minutes = Math.floor((time % 3600) / 60);
    const seconds = time % 60;

    const pad = (num) => num.toString().padStart(2, '0');

    if (showHours) {
      return `${pad(hours)}:${pad(minutes)}:${pad(seconds)}`;
    }
    return `${pad(minutes)}:${pad(seconds)}`;
  }, [showHours]);

  useEffect(() => {
    let timer;
    if (isRunning && timeLeft > 0) {
      timer = setInterval(() => {
        setTimeLeft((prev) => {
          if (prev <= 1) {
            clearInterval(timer);
            onTimeUp?.();
            return 0;
          }
          return prev - 1;
        });
      }, 1000);
    }
    return () => clearInterval(timer);
  }, [isRunning, timeLeft, onTimeUp]);

  const getColorStyle = () => {
    if (timeLeft <= 60) return styles.danger;
    if (timeLeft <= 300) return styles.warning;
    return styles.normal;
  };

  return (
    <View style={styles.container} testID={testID}>
      <Text style={[styles.time, getColorStyle()]} testID="timer-text">
        {formatTime(timeLeft)}
      </Text>
    </View>
  );
};

Timer.propTypes = {
  duration: PropTypes.number.isRequired,
  onTimeUp: PropTypes.func,
  isRunning: PropTypes.bool,
  showHours: PropTypes.bool,
  testID: PropTypes.string
};

const styles = StyleSheet.create({
  container: {
    padding: 8,
    borderRadius: 8,
    backgroundColor: '#FFFFFF',
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 1 },
    shadowOpacity: 0.1,
    shadowRadius: 2,
    elevation: 2,
  },
  time: {
    fontSize: 20,
    fontWeight: '600',
    textAlign: 'center',
  },
  normal: {
    color: '#2D3748',
  },
  warning: {
    color: '#C05621',
  },
  danger: {
    color: '#C53030',
  },
});

export default Timer;
