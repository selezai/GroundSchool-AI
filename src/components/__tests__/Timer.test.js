import React from 'react';
import renderer, { act } from 'react-test-renderer';
import Timer from '../Timer';

jest.useFakeTimers();

describe('Timer Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    jest.clearAllTimers();
  });

  it('renders correctly with minutes and seconds', () => {
    let tree;
    act(() => {
      tree = renderer.create(
        <Timer duration={300} onTimeUp={() => {}} />
      ).toJSON();
    });
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with hours format', () => {
    let tree;
    act(() => {
      tree = renderer.create(
        <Timer duration={3600} showHours={true} onTimeUp={() => {}} />
      ).toJSON();
    });
    expect(tree).toMatchSnapshot();
  });

  it('shows warning color when time is under 5 minutes', () => {
    let tree;
    act(() => {
      tree = renderer.create(
        <Timer duration={299} onTimeUp={() => {}} />
      ).toJSON();
    });
    expect(tree).toMatchSnapshot();
  });

  it('shows danger color when time is under 1 minute', () => {
    let tree;
    act(() => {
      tree = renderer.create(
        <Timer duration={59} onTimeUp={() => {}} />
      ).toJSON();
    });
    expect(tree).toMatchSnapshot();
  });

  it('calls onTimeUp when timer reaches zero', () => {
    const onTimeUp = jest.fn();
    act(() => {
      renderer.create(
        <Timer duration={2} onTimeUp={onTimeUp} />
      );
    });

    // Fast-forward until all timers have been executed
    act(() => {
      jest.advanceTimersByTime(2000);
    });

    expect(onTimeUp).toHaveBeenCalled();
  });

  it('stops counting when isRunning is false', () => {
    let component;
    act(() => {
      component = renderer.create(
        <Timer duration={10} isRunning={false} onTimeUp={() => {}} />
      );
    });

    const initialTime = component.root.findByProps({ testID: 'timer-text' }).props.children;

    act(() => {
      jest.advanceTimersByTime(1000);
    });

    const timeAfterDelay = component.root.findByProps({ testID: 'timer-text' }).props.children;
    expect(timeAfterDelay).toBe(initialTime);
  });
});
