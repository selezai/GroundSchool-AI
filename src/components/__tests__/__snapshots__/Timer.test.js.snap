// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer Component renders correctly with hours format 1`] = `null`;

exports[`Timer Component renders correctly with minutes and seconds 1`] = `null`;

exports[`Timer Component shows danger color when time is under 1 minute 1`] = `null`;

exports[`Timer Component shows warning color when time is under 5 minutes 1`] = `null`;
